File: packages/sdk/src/payments.ts
import { QelosSDKOptions } from './types';
import BaseSDK from './base-sdk';
import { IPlan, ISubscription, IInvoice, ICoupon, BillingCycle } from '@qelos/global-types';

export interface CheckoutRequest {
  /** Required when not providing `subscriptionId`. Must be a non-dynamic plan. */
  planId?: string;
  /** Required when not providing `subscriptionId`. */
  billingCycle?: BillingCycle;
  couponCode?: string;
  successUrl?: string;
  cancelUrl?: string;
  /**
   * Pre-created subscription ID. Required for dynamic plans — the admin must set
   * `dynamicAmount` on the subscription before this checkout call will succeed.
   * When provided, `planId` and `billingCycle` are ignored.
   */
  subscriptionId?: string;
}

export interface CheckoutResponse {
  subscriptionId: string;
  checkoutUrl?: string;
  clientToken?: string;
}

export interface ValidateCouponResponse extends ICoupon {}

export default class QlPayments extends BaseSDK {
  private relativePath = '/api';

  constructor(options: QelosSDKOptions) {
    super(options);
  }

  getPlans(query?: { isActive?: boolean }) {
    const qs = query ? `?${new URLSearchParams(query as any)}` : '';
    return this.callJsonApi<IPlan[]>(`${this.relativePath}/plans/public${qs}`);
  }

  getPlan(planId: string) {
    return this.callJsonApi<IPlan>(`${this.relativePath}/plans/${planId}`);
  }

  /**
   * Creates a pending subscription for the authenticated user without initiating
   * payment. Use this as the first step for any plan, and the only step before
   * `checkout()` for dynamic plans (the admin must call `setSubscriptionDynamicAmount`
   * before the user can complete checkout).
   */
  subscribeToPlan(planId: string, billingCycle: BillingCycle, couponCode?: string) {
    return this.callJsonApi<ISubscription>(
      `${this.relativePath}/subscriptions`,
      {
        method: 'post',
        headers: { 'content-type': 'application/json' },
        body: JSON.stringify({ planId, billingCycle, couponCode }),
      },
    );
  }

  /**
   * Initiates a provider checkout session.
   *
   * **Static plans**: pass `planId` + `billingCycle` (a subscription is created
   * inline).
   *
   * **Dynamic plans**: the user must first call `subscribeToPlan()` to create a
   * pending subscription, then an admin must set the dynamic amount via
   * `setSubscriptionDynamicAmount()`. Only then should this method be called with
   * `subscriptionId` pointing to that pending subscription.
   */
  checkout(params: CheckoutRequest) {
    return this.callJsonApi<CheckoutResponse>(
      `${this.relativePath}/checkout`,
      {
        method: 'post',
        headers: { 'content-type': 'application/json' },
        body: JSON.stringify(params),
      },
    );
  }

  getMySubscription() {
    return this.callJsonApi<ISubscription>(`${this.relativePath}/subscriptions/me`);
  }

  cancelSubscription(subscriptionId: string) {
    return this.callJsonApi<ISubscription>(
      `${this.relativePath}/subscriptions/${subscriptionId}/cancel`,
      { method: 'put' },
    );
  }

  getInvoices(query?: { status?: string }) {
    const qs = query ? `?${new URLSearchParams(query as any)}` : '';
    return this.callJsonApi<IInvoice[]>(`${this.relativePath}/invoices${qs}`);
  }

  getInvoice(invoiceId: string) {
    return this.callJsonApi<IInvoice>(`${this.relativePath}/invoices/${invoiceId}`);
  }

  validateCoupon(code: string, planId?: string) {
    return this.callJsonApi<ValidateCouponResponse>(
      `${this.relativePath}/coupons/validate`,
      {
        method: 'post',
        headers: { 'content-type': 'application/json' },
        body: JSON.stringify({ code, planId }),
      },
    );
  }
}

File: packages/python-sdk/qelos_sdk/payments.py
from __future__ import annotations

import json
from typing import Any, Dict, List, Optional
from urllib.parse import urlencode

from .base_sdk import BaseSDK
from .types import QelosSDKOptions


class QlPayments(BaseSDK):
    """User-facing payment operations: plans, checkout, subscriptions, invoices, and coupons."""

    _relative_path = "/api"

    def __init__(self, options: QelosSDKOptions) -> None:
        super().__init__(options)

    async def get_plans(self, query: Optional[Dict[str, Any]] = None) -> List[Dict[str, Any]]:
        """Get public plans."""
        qs = "?" + urlencode(query) if query else ""
        return await self.call_json_api(f"{self._relative_path}/plans/public{qs}")

    async def get_plan(self, plan_id: str) -> Dict[str, Any]:
        """Get a plan by ID."""
        return await self.call_json_api(f"{self._relative_path}/plans/{plan_id}")

    async def checkout(self, params: Dict[str, Any]) -> Dict[str, Any]:
        """Create a checkout session.

        Args:
            params: Dict with ``planId``, ``billingCycle``, and optional
                ``couponCode``, ``successUrl``, ``cancelUrl``.

        Returns:
            Dict with ``subscriptionId`` and optional ``checkoutUrl``, ``clientToken``.
        """
        return await self.call_json_api(
            f"{self._relative_path}/checkout",
            method="POST",
            headers={"content-type": "application/json"},
            body=json.dumps(params),
        )

    async def get_my_subscription(self) -> Dict[str, Any]:
        """Get the current user's subscription."""
        return await self.call_json_api(f"{self._relative_path}/subscriptions/me")

    async def cancel_subscription(self, subscription_id: str) -> Dict[str, Any]:
        """Cancel a subscription."""
        return await self.call_json_api(
            f"{self._relative_path}/subscriptions/{subscription_id}/cancel",
            method="PUT",
        )

    async def get_invoices(self, query: Optional[Dict[str, Any]] = None) -> List[Dict[str, Any]]:
        """List invoices."""
        qs = "?" + urlencode(query) if query else ""
        return await self.call_json_api(f"{self._relative_path}/invoices{qs}")

    async def get_invoice(self, invoice_id: str) -> Dict[str, Any]:
        """Get an invoice by ID."""
        return await self.call_json_api(f"{self._relative_path}/invoices/{invoice_id}")

    async def validate_coupon(self, code: str, plan_id: Optional[str] = None) -> Dict[str, Any]:
        """Validate a coupon code."""
        return await self.call_json_api(
            f"{self._relative_path}/coupons/validate",
            method="POST",
            headers={"content-type": "application/json"},
            body=json.dumps({"code": code, "planId": plan_id}),
        )
